homeassistant/auth/providers/command_line.py

```diff
@@ -0,0 +1,163 @@
+"""Auth provider that validates credentials via an external command."""
+
+from typing import Any, Dict, Optional, cast
+
+import asyncio.subprocess
+import collections
+import logging
+import os
+
+import voluptuous as vol
+
+from homeassistant.exceptions import HomeAssistantError
+
+from . import AuthProvider, AUTH_PROVIDER_SCHEMA, AUTH_PROVIDERS, LoginFlow
+from ..models import Credentials, UserMeta
+
+
+CONF_COMMAND = "command"
+CONF_ARGS = "args"
+CONF_META = "meta"
+
+CONFIG_SCHEMA = AUTH_PROVIDER_SCHEMA.extend({
+    vol.Required(CONF_COMMAND): vol.All(
+        str,
+        os.path.normpath,
+        msg="must be an absolute path"
+    ),
+    vol.Optional(CONF_ARGS, default=None): vol.Any(vol.DefaultTo(list), [str]),
+    vol.Optional(CONF_META, default=False): bool,
+}, extra=vol.PREVENT_EXTRA)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class InvalidAuthError(HomeAssistantError):
+    """Raised when authentication with given credentials fails."""
+
+
+@AUTH_PROVIDERS.register("command_line")
+class CommandLineAuthProvider(AuthProvider):
+    """Auth provider validating credentials by calling a command."""
+
+    DEFAULT_TITLE = "Command Line Authentication"
+
+    # which keys to accept from a program's stdout
+    ALLOWED_META_KEYS = ("name",)
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        """Extend parent's __init__.
+
+        Adds self._user_meta dictionary to hold the user-specific
+        attributes provided by external programs.
+        """
+        super().__init__(*args, **kwargs)
+        self._user_meta = {}  # type: Dict[str, Dict[str, Any]]
+
+    async def async_login_flow(self, context: Optional[dict]) -> LoginFlow:
+        """Return a flow to login."""
+        return CommandLineLoginFlow(self)
+
+    async def async_validate_login(self, username: str, password: str) -> None:
+        """Validate a username and password."""
+        env = {
+            "username": username,
+            "password": password,
+        }
+        try:
+            # pylint: disable=no-member
+            process = await asyncio.subprocess.create_subprocess_exec(
+                self.config[CONF_COMMAND], *self.config[CONF_ARGS],
+                env=env,
+                stdout=asyncio.subprocess.PIPE
+                if self.config[CONF_META] else None,
+            )
+            stdout = (await process.communicate())[0]
```

Comment:
You want each line of stderr to be logged via the logging system in case authentication fails? At which level?

What should happen with stderr when authentication succeeds? I ask because it's sometimes used for status messages that are no errors.

Reply:
Right now, stderr always goes to stdout of the HA process.

Reply:
I see, stderr was ignored in `create_subprocess_exec`, so this line can be changed a little bit 
```python
stdout, _ = await process.communicate()
```